### ISSUE TITLE
Bug/Feature: Comments on first line

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#829]
This might involve both a bug and a feature.

Feature: I need to be able to set a comment for the whole sequence, something like:

"Focus on Spin the Top"

This was brought up by Travis at Callerlab, but I've wanted this as well.

Now, I could stick this on the first call as a prefix or suffix comment (which is lame), but it really wants to be a standalone comment that has no call associated with it.

So, either I need to:

- add support for standalone comments (this allows N standalone comments), OR 
- I could just extend the on-disk format to have a DESCRIPTION field, e.g. `#DESC=Focus on Spin the Top#`, and then stick this into a Label that is right above the CurrentSequence widget.  This solution allows only one "whole sequence" standalone comment, and it requires a special dialog to be able to add it after-the-fact.

**Comment by mpogue2** [comment]
**SD sequence that inserts a comment onto the first line:**
- heads start;insert a comment;foo;square thru 4

result: "HEADS { foo } square thru 4"

**SD sequence that does NOT work:**
- insert a comment;

result: beep (needs a startup command like HEADS start)

standalone comments, e.g. "(comment with no call)" are not accepted
inline comments, e.g. "scoot back and ( girls ) roll" are not accepted

**Comment by mpogue2** [comment]
### TODAY
Today, this translation is done on input:
`"Square Thru 4 ( suffix comment )" --> "insert a comment" "(| suffix comment |)" "Square Thru 4"`
and on output:
`"{| suffix comment |} Square Thru 4" --> "Square Thru 4 ( suffix comment )"`

Today, this translation is also done on input:
`"( prefix comment ) Square Thru 4" --> "insert a comment" "( prefix comment )" "Square Thru 4"`
and on output:
`"{ prefix comment } Square Thru 4" --> "( prefix comment ) Square Thru 4"`

So, round trip, the user always sees what they typed in.  And, what's stored on disk is what the user typed in (normalized, but essentially the same).

### TODO TO GET THE FIRST LINE COMMENTS WORKING
Because the first line is "special" in SD, that implies that we must do these:
`"HEADS Square Thru 4 ( suffix comment )" --> "heads start" "insert a comment" "(| suffix comment |)" "Square Thru 4"`
and on output:
`"HEADS {| suffix comment |} Square Thru 4" --> "HEADS Square Thru 4 ( suffix comment )"`

and:
`"( prefix comment ) HEADS Square Thru 4" --> "heads start" "insert a comment" "( prefix comment )" "Square Thru 4"`
and on output:
`"HEADS { prefix comment } Square Thru 4" --> "( prefix comment ) HEADS Square Thru 4"`

Note that "Heads start" is NOT present in the SD output stream that we use to populate the CurrentSequence table.  That gets inserted at input time only, because SD requires it for the initial command.

**Comment by mpogue2** [comment]
A different way to go might be this:

- If user enters some call with some comments (anywhere), e.g. `"( prefix ) Heads ( inline1 ) square ( inline 2 ) thru 4 ( suffix )"` then

A) if FIRST LINE, we submit this to SD as `"heads start" "insert a comment" "( prefix ) Heads ( inline1 ) square ( inline 2 ) thru 4 ( suffix )" "square thru 4"`
- and on output, SD will give us "{( prefix ) Heads ( inline1 ) square ( inline 2 ) thru 4 ( suffix )} heads square thru 4"
- and we will just use "( prefix ) Heads ( inline1 ) square ( inline 2 ) thru 4 ( suffix )" for the user display

B) else if NOT FIRST LINE we enter this into SD as `"insert a comment" "( prefix ) Heads ( inline1 ) square ( inline 2 ) thru 4 ( suffix )" "HEADS square thru 4"`
- and on output, SD will give us "{( prefix ) Heads ( inline1 ) square ( inline 2 ) thru 4 ( suffix )} heads square thru 4"
- which will will translate to "( prefix ) Heads ( inline1 ) square ( inline 2 ) thru 4 ( suffix )" for display

So, basically what is in the { } will be what the user sees, and we ALWAYS submit that to SD as a comment before each line, with a bit of special processing for the HEADS START on the first line.  If the call without the parenthesized comments is NULL, we don't submit anything to SD.

PRO:
I think this would allow us to implement prefix, inline, suffix, and standalone comments in one fell swoop.

CON:
We'd have to take the hit for changing "Add Comment" and "Delete Comment" to "Modify Call", and the yucky error processing that it implies.  We could just leave Add/Delete Comment there as is, even though it's "suffix comments only".

**Comment by mpogue2** [comment]
OK, after some thought, I think I'm gonna go with this:

1. The "different" option described above.  This is quite a change from what I did before (a major rewrite, most likely), but the advantages are that all types of comments (except standalone) can be supported.  I plan to implement two functions that map directly to the two transformations described above, something like: `UserInputToSDCommands()` and `SDOutputToUserOutput()`.  
2. I will add an editable text field for the Title of the sequence.  It will be editable only when the sequence is UNLOCKed.  It will be saved not as a standalone comment, but using the metadata style: "#TITLE=<title here>#".

The result will be support for both a TITLE for each sequence, e.g. "Left Fan the Top practice", and for 3 kinds of comments, e.g. "( prefix comment ) call", "call ( suffix comment )", and "part1 (inline) part2".

**Comment by mpogue2** [comment]
OK, I implemented this new "different" option, and it kinda does some things better and some things worse:

**Better:**
- I can put a comment on line 1 now (was not possible before)
- I can put comments anywhere in the line and they are preserved, e.g. "(comment 1) Heads (comment 2) Square Thru (comment 3)" 

**Not so good:**
- I have lost round-tripping for the following reason:
   Now the original user-typed line is preserved as a comment, and displayed as the correct line.  BUT, this is missing all the transformations I applied to it, so "Heads square thru", which is NOT accepted by SD natively, is transformed to "Heads square thru 4", which is, but what is shown to the user and recorded in the file is "Heads square thru".  This impedance mismatch occurs only for lines with comments.  Lines without comments are transformed, then shown and saved as the transformed version.
- Sometime an input line will result in 2 output lines, which are just concatenated right now to the Current Sequence table, e.g. "Follow your Neighbor and (girls) Spread".   This is transformed to:  "{Follow your Neighbor and (girls) spread} [Follow your neighbor] and spread", which is then way long.  The two line SD output then results in the user seeing: "Follow your [Follow your neighbor] and spread" or something similar.  Yuck.
- I still can't do "( standalone comments )".

The root of the problem here is that the transformed call does not easily map to the input line, when comments can be placed anywhere, for example:

"Heads ( comment 1 ) Square (comment 2) thru"
maps internally to 
"{Heads ( comment 1 ) Square (comment 2) thru} heads square thru 4", which pops out of SD on the output side, but then there's no easy way to get comments 1 & 2 back into the heads square thru 4 part.

I gotta think about this some more...

**Comment by mpogue2** [comment]
Rewriting to use the old way, which allows prefix and suffix comments, but not inline or standalone comments.

Bugs:
1) highlighting bug:  if "swing thru ( comment )" is highlighted, it is in the file as "#HIGHLIGHT=swing thru (hello)#" and it should be "#HIGHLIGHT=swing thru#"
2) sequence: "Square thru 4 (suffix)" doesn't have "HEADS added". This is because what we send to SD is "HEADS { |suffix| } square thru 4, which is not parsed to theCall as "HEADS Square thru 4".

**Comment by mpogue2** [comment]
Bug: standalone comments are not accepted, but when saved the sequence shows up as "{ |no Matches)|}".
UNDO can get rid of it, but it should not be there at all.

**Comment by mpogue2** [comment]
Bug: autocomplete doesn't work if a prefix comment is used, e.g. ( foo ) ferris <TAB>

**Comment by mpogue2** [comment]
All except the last bug listed was fixed by: 680e69340ea7877e784e247f152a64b764007ba5 .
I'll open a new issue to track that one.
